sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/StubbedKinesisAsyncClient.java

```diff
@@ -79,7 +80,7 @@ public CompletableFuture<Void> subscribeToShard(
     subscribeRequestsSeen.add(req);
     Deque<StubbedSdkPublisher> publishers =
         checkNotNull(stubbedPublishers.get(req.shardId()), "Not stubbed");
-    StubbedSdkPublisher publisher = checkNotNull(publishers.poll(), "No more stubs");
+    StubbedSdkPublisher publisher = Objects.firstNonNull(publishers.poll(), new NoopSdkPublisher());
```

Comment:
I think it's best to insert a noop publisher at the very end. That way tests can be written in a more stable way as timings will be hard to get right otherwise.

Reply:
Thanks, this is a good idea. I thought throwing an NPE here and having it more "harsh" would be better for more strict testing, but looks like it makes tests more complicated instead. Let's keep this `NoopSdkPublisher`